### DISCUSSION TITLE
Extremely slow debugging

### DISCUSSION THREAD

**Discussion by mcergun** [polarfire-soc/polarfire-soc-documentation#342] [How to]
I've been fiddling with the Icicle Kit on and off for at least a year now. The debugging has always been slow for me, to the point that it can't be considered useable. 

When stepping through the code, the next line only comes up after several seconds. The time it takes is generally related to the number of variables/registers it parses, or the number of cores it follows. I can literally see each variable being read/parsed when watching the IDE.

The problem can be replicated even with the basic bare-metal examples supplied on your GitHub repositories. The problem does not happen when debugging through Renode emulation.

I've changed SoftConsole versions but that doesn't change the result. I need a starting point where I can pinpoint the real problem.

**Comment by hughbreslin** [comment]
I would recommend trying your debug with an Olimex ARM-USB-TINY-H, just note this isn’t pin compatible so you will need to use jumper cables to connect it.

**Reply by mcergun** [reply]
How can we make sure embedded flash pro is the source of the slowness before buying new hw to try?

Is this a known limitation of the embedded debugger?

**Reply by hughbreslin** [reply]
I’ve tried it myself and can say the olimex is faster for embedded software debug.

**Reply by mcergun** [reply]
I see, but is the efp6b slow for your use? Like a simple arithmetic operation stepping takes 4 to 5 seconds on my end. I personally don't need blazing fast right now, acceptable fast is enough.

**Reply by hughbreslin** [reply]
Yes it is slow for embedded debug - my workaround is to use breakpoints and jump multiple instructions in one go instead of single stepping where possible as this is a faster approach.

**Reply by grg-haas** [reply]
I'll note that I have a full flash pro 6 (not the one embedded on the board) and it is just as slow at debugging. Thanks for the hint on the Olimex part!

**Comment by IamMattM** [comment]
@hughbreslin 
Thanks.
Our custom board will expose the JTAG port at +1.8V
For this reason we are considering using the Olimex debugger that also supports the 1.8V IO levels : <ARM-USB-OCD-H>

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/dd1aff83-51a8-47c1-8555-450d96ee90e5)

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/edffbf01-fbc3-43ad-adb0-8a87b0ae8b96)


Should softconsole/EclipseIDE be able to "just" attach to this debugger with different usb pid ? 

I had noted down that the polarfire-soc should support JTAG_CLK up to 25MHz and the Olimex debugger mentions "RTCK adaptive JTAG clock up to 30Mhz" - which would surpass the MPFS clocking capabilities so we should be able to maximise debugging throughput.
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/f2986210-4dbf-4f81-965d-f8d22d38bd0e)



Saying that, isn't the slowliness of the FlashPro6 not simply coming down to a default frequency of 4MHz when it should be possible to "force select" the JTAG clock up to 20MHz with FP6 (and actualy up to the full 25MHz with the earlier FP5) ?

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/921f5f9e-8198-4424-8751-2d2655a44111)

Let me know before making a decision on whether we need additional debugger for faster debug.

**Reply by grg-haas** [reply]
I will note that I did not know about this programmer settings menu in Libero, and forcing the TCK frequency to 20MHz did significantly speed up my debugging

**Reply by IamMattM** [reply]
@grg-haas 
"forcing the TCK frequency to 20MHz did significantly speed up my debugging" - 

Great. Thank you for trying this out and confirming the debugging speed improvment.

We might not need to acquire an additional debugger after all.

**Reply by mcergun** [reply]
How do you force tclk? 
I have seen the setting on fpexpress but assumed it was only for the flashing session, not permanent (i.e. for soft console debugging)

**Reply by IamMattM** [reply]
Is there not anywhere in the softconsole/eclipse IDE GUI / debugger config TAB (or perhaps the openocd config file called upon) where the JTAG clock can be changed from default  ?
```
# Variable name: JTAG_KHZ
# Optional/required: Optional 
# Allowed values: 1000, 2000, 3000, 4000, 5000 (FlashPro4/FlashPro5 only), 
# 6000, 10000 (FlashPro5 only)
# Default value: 6000 (set in interface/microsemi-flashpro.cfg)
# Description: JTAG TCK clock speed in kHz (e.g. 1000 = 1000kHz = 1MHz).
# Example: --command "set JTAG_KHZ 4000"
```

Not tried it myself yet but I would imagine this is the place where one could request a 20000 KHz JTAG clock.
Seems to tie up with a different post showing a default 6MHz JTAG clock:

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/4e2b03a4-1639-47bd-9590-710ac38e2f61)

**Comment by IamMattM** [comment]
@mcergun

If softconsole is installed for you under Windows, you will find the openocd jtag clock parameter used under:

C:\Microchip\SoftConsole-v2022.2-RISC-V-747\openocd\share\openocd\scripts\interface\microsemi-flashpro.cfg

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/69704e7d-bae3-4bf5-960b-e76664165e4a)

**Reply by IamMattM** [reply]
Since the "exists" condition check is there in the <microsemi-flashpro.cfg> file, I suppose one could instead add the --command line highlighted below and leave the default 6MHz for compatibility with earlier FP3/4 and it should do the trick ?

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/c34272a2-e7be-4a8e-b255-edb3083983f9)

**Reply by mcergun** [reply]
This makes sense. Will be trying this as soon as I get the chance.

**Reply by mcergun** [reply]
I have the fp5 which supports 6, 10, 15mhz clk speeds.
Sadly, setting the value above 6mhz does not work for debugging. Gdb says it can't detect device info and fails.

**Reply by IamMattM** [reply]
Hmm...
Everything in the softconsole help/documentation and the microsemi-flashpro.cfg seems to indicate that it should work. (perhaps your standalone FP5 programmer needs a firmware update ?)

What about using the embedded eFP6revB as your initial post suggests you were using ?

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/4c41c4d3-dcc0-4b25-b7f5-9a525da93a81)

**Reply by IamMattM** [reply]
Interestingly the mentionned alternate OLIMEX openocd debugger target .cfg referred to here:
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/62ff0537-29b0-4e16-9da7-207c16e48f1d)

is no longer showing up in the installation folder so if one was to use the recommended Olimex, I imagine that we would need to point to using the file but it is missing...

openocd/share/openocd/scripts/target/microsemi-olimex-tiny-h-riscv.cfg
@hughbreslin  : Would you happen to have a copy of the "microsemi-olimex-tiny-h-riscv.cfg" file should one purchase the Olimex Tiny-h debugger to accelerate development (as it seeem the FP6 is limited to just 6MHz JTAG_CLK ?).

**Reply by IamMattM** [reply]
We have been able to try the embeddedFP6revb at 20MHz using "--command "set JTAG_KHZ 20000" and it is working.

However with the standalone FP6 the maximum working JTAG_CLK has been found to be 11MHz only...? We tried a shorter ribbon cable of 6cm (compared to 12cm of the provided one) and this has *not* helped in achieving >11MHz.

This means that we might recourse to using a third party open-ocd debugger to achieve the >20MHz JTAG_CLK to make the software debugging speeds acceptable.
 
@hughbreslin : Sorry to chase you up on this once more but would you happen to have those *olimex*.cfg files mentionned above ?

Would you happen to have an indication as to why the official Microchip standalone debugger is nowhere near as good as the embedded implementtion on the Icicle ?

UPDATE:
It would appear that there was a slight improvment with the FP6 after using FlashPro Express in developer mode and reflashing the debugger to v4.2 : JTAG_CLK=12MHz now working. 
Still not the 20MHz of the eFP6 *revb*

Is there a hw difference between the eFP6revB and the standalone FP6 that would explain the limited supported JTAG_CLK of the standalone FP6 ?

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/29542928/cc6207e0-b9f6-4c26-9b6c-e1117133d80b)

**Reply by IamMattM** [reply]
@mcergun
Could you try to update your stadalone FP5 using FPExpress-developer mode and try out softconsole with the "--command "set JTAG_KHZ 20000"

Just curious as to whether the FP5 might actually be working faster than the new FP6 through Softconsole program.
